### PR TITLE
Two step otp

### LINKED ISSUE #667: Use `standalone: false` or embed QR codes as an image
The QR code rendered during 2FA setup currently has `standalone: true` set as the default. The `standalone` option might be a little misleading in this case. As `standalone: true` will not only output wrapping `sag` tags, but also an XML declaration, i.e. "standalone" in this case refers to using the SVG as a standalone document. That’s not what we're doing here though -- we’re rendering the SVG as part of an HTML document.

There’s two possible solutions:

**1) Use `standalone: false`.**
We’d then need to wrap the generated elements in an `svg` tag ourselves. This would also require that we calculate the intrinsic width of the QR code ourselves, which is a little annoying, but it’s also easy to do as `rqrcode` exposes the number of modules and their respective size.

**2) Use `standalone: true` and encode the document as a data URI**
The other option is to keep using `standalone: true`, but instead of embedding the QR code as an inline SVG, we could encode it as a data URI and them embed it as an `img`:

```ruby
    def qr_code_data_uri
      "data:image/svg+xml,#{URI::encode(qr_code.as_svg)}"
    end
```

### PR DESCRIPTION
**Pull request by tillprochaska**
This PR is based on #949. For review, have a look at that PR first. Merge #949 first.

- [ ] ~As this implementation stores wether the OTP for the signed in user has already been verified in a session variable, right now, users would be prompted to confirm the OTP every time they start a new browser session. This is out of sync with the time the login is remembered (24 hours). I think it might be acceptable to simply set the expiration duration of the `remember_token` to the browser session as well.~

Closes #667

https://user-images.githubusercontent.com/1512805/132007189-28dd3d28-2e14-426a-bd3c-bc0753dfeb8e.mov

### CODE REVIEW

**Comment by FrauCsu** [comment]
Looks great @tillprochaska ! #improvement #hiphiphooray

**Comment by drjakob** [comment]
Abmelden sollte oben rechts über Suchen stehen

**Comment by drjakob** [comment]
"In drei Schritten richtest du die Anmeldung in zwei Schritten ein" ist vollkommen wirr.  "Fast fertig! Richte noch die Absicherung über Zwei-Faktor-Authentifizierung ein." (Ich würde den richtigen Begriff nehmen.)

**Comment by tillprochaska** [comment]
@roschaefer @phoeinx I updated the implementation as outlined 
3125985. I’d still like to add additional request specs for the custom behavior in session and passwords controllers, but apart from that, I think it’s ready for review

**Review by roschaefer — COMMENTED** [review]
I have to catch a bus soon, however, I gave this a review and it works and it looks like the issue with being half-signed-in has been addressed.

Nevertheless, I see some complexity because many method calls are scattered across different controllers. E.g. `sessions` controller calls `sign_in` which is defined in `application_controller` which in turn calls `super`.  I looked into the source code of clearance and I think we can refactor this PR to re-use more of the clearance-maintained source code. E.g. some of our `create` routes look very similar. Maybe we can do a pair-programming together @tillprochaska ? Since this PR touches security, I think it's worth it.

**Comment by tillprochaska** [comment]
> Maybe we can do a pair-programming together @tillprochaska ? Since this PR touches security, I think it's worth it.

Yes, let’s do that!

Although, I already tried to duplicate little code as possible. Unfortunately, clearance doesn't really provide any methods to hook into the default behavior (except from where to redirect). That’s why I decided to go for the initial implementation. I’ve opened an issue in the Clearance repo asking if they’d accept a PR changing this (e.g. to customize what to do after a successful password reset).

By the way, I also evaluated a few different auth libraries, to see if any of them fits our needs better. Currently, well-maintained libraries include Devise, Authlogic, Clearance, and Rodauth. The three former all do not support 2FA out of the box. There are a few Devise plugins that provide 2FA. However, I think all of them have the same bad UX we currently have (i.e. they ask for email, password *and* OTP in one step).

The latter has support for 2FA built-in. However, it also has a few drawbacks: It’s an extensive library, written in Roda (Framework) and Sequel (ORM). It has adapter for Rails and that works quite well (I tried that out), possibly even without the need for database migrations. More importantly, it does also implement 2FA in the same way I did initially (i.e. the "semi-signed-in" state).

**Review by roschaefer — APPROVED** [review]
complexity

I'm going to approve this, as I don't want this to be blocked. Feel free to merge and create a follow-up issue to refactor the four controllers `ApplicationController`, `OtpSetupController`, `OtpAuthController` and `SessionsController`.

We can also do it together *before* you merge it. I would be totally up for it.

**Comment by tillprochaska** [comment]
Yes, let's do it before we merge!

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -2,39 +2,38 @@
...
 
-  def default_url_options(options = {})
-    return options.merge(locale: I18n.locale) unless I18n.locale == I18n.default_locale
-
-    options
-  end
-
-  def locale_params
-    params.permit(:locale)
+  def require_otp_setup
```
[the remark this answers is not in the document]
Comment (tillprochaska) [reply app/controllers/application_controller.rb:13]: I like the mixin/concern solution, thought about that as well.

**Comment by tillprochaska** [comment]
As discussed with @roschaefer, I’m merging this. We’re going to do the remaining refactoring in a separate PR.

**Pull request merged**